Allow setting HiPS CORS and credential options

pmatsson opened the pull request:
Hi,

This PR adds the CORS and Credential options when loading a HiPS.

**Use case:**

- We want to be able to display a proprietary HiPS for a select group of users.
- The HiPS is accessible over the internet but is behind some form of authentication.
- For this we need to be able to set the CORS and Credentials options when requesting Aladin to load a HiPS.

pmatsson committed:
Allow setting HiPS CORS and credential options

bmatthieu3 commented on src/js/HiPS.js:
```diff
@@ -210,6 +210,8 @@ export let HiPS = (function () {
         this.options = options;
         this.name = (options && options.name) || id;
         this.startUrl = options.startUrl;
+        this.requestMode = options && options.requestMode;
```
Could you add some documentation in the HiPSOptions for this new option to know what it is and the possible values it can have ?

pmatsson replied:
Added.

bmatthieu3 commented on src/js/HiPS.js:
```diff
@@ -210,6 +210,8 @@ export let HiPS = (function () {
         this.name = (options && options.name) || id;
         this.startUrl = options.startUrl;
+        this.requestMode = options && options.requestMode;
+        this.requestCredentials = options && options.requestCredentials;
```
Same for this one

bmatthieu3 commented on src/js/HiPS.js:
```diff
@@ -902,6 +904,8 @@ export let HiPS = (function () {
                 hipsCubeDepth: self.cubeDepth,
                 isPlanetaryBody: self.isPlanetaryBody(),
                 hipsBody: self.hipsBody,
+                requestCredentials: self.requestCredentials || 'omit',
```
Is it possible to set directly `this.requestCredentials` to `omit` in the constructor (if no value is given) instead of here.

pmatsson replied:
Fixed.

bmatthieu3 commented on src/js/HiPS.js:
```diff
@@ -902,6 +904,8 @@ export let HiPS = (function () {
                 isPlanetaryBody: self.isPlanetaryBody(),
                 hipsBody: self.hipsBody,
+                requestCredentials: self.requestCredentials || 'omit',
+                requestMode: self.requestMode || 'cors',
```
Same

bmatthieu3 commented on src/core/src/downloader/query.rs:
```diff
@@ -11,26 +11,31 @@ pub type QueryId = String;
 
 use al_core::image::format::ImageFormatType;
 
-#[derive(Eq, Hash, PartialEq, Clone)]
```
Did not you have a compilation error after removing it ? Or maybe it was unecessary for Tile to be hashable and in that case that is a good thing to remove it.

pmatsson replied:
No compilation errors, from what I could tell it is not needed. We would have to implement the Hash trait for RequestMode and RequestCredentials if we would like to keep it.

bmatthieu3 commented on src/core/src/downloader/request/allsky.rs:
```diff
@@ -132,7 +134,8 @@ impl From<query::Allsky> for AllskyRequest {
                     let mut opts = RequestInit::new();
                     opts.method("GET");
-                    opts.mode(RequestMode::Cors);
+                    opts.mode(mode);
```
line 122 of that file: I think we might need to extend `query_image` with the new credentials variable (same as what you did for Tile::query_html_image). And actually, this Allsky::query_image method could use the query_html_image more generic way. I think I will have to refactorize that one time. For the moment you can just pass the credentials as well here I think.

pmatsson replied:
Fixed.

bmatthieu3 reviewed (changes requested):
Hi @pmatsson thanks for this PR. We are interested in merging it. I did a review.
I think the Moc.fits located at the root of the HiPS folder should be treated as well. This moc request case is managed in `request/moc.rs`.
Also did you tested on your side this PR ? Maybe on a local HiPS you put under credentials ? Thank you very much

pmatsson committed:
Support credentials on allsky queries, adding docs and cleaning up

pmatsson committed:
Support credentials on MOC queries

esdc-esac-esa-int commented:
Hi, thanks for the review. I've updated the code with your suggestions and removed some dead code that I had added. 

We have tested this code with some HiPS behind basic auth on a remote server, and it works perfectly for our use case. The remote server must allow CORS from the origin of the request (allowing * will result in CORS issues).

pmatsson commented:
Hi, thanks for the review. I've updated the code with your suggestions and removed some dead code that I had added.

We have tested this code with some HiPS behind basic auth on a remote server, and it works perfectly for our use case. The remote server must allow CORS from the origin of the request (allowing * will result in CORS issues).

bmatthieu3 commented:
Thank you @pmatsson, this is much clearer with the docs. May I ask if you have a javascript example on how to request for such a private HiPS ? Thanks

pmatsson commented:
> Thank you @pmatsson, this is much clearer with the docs. May I ask if you have a javascript example on how to request for such a private HiPS ? Thanks

For example:
`aladin.addNewImageLayer(A.HiPS('https://some-private-hips.com/myhips', {name: 'myHips', requestCredentials: 'include', requestMode: 'cors'}))`

If the user has authenticated, the auth cookie will be included in the request, thus authenticating the user. If not the user will be presented by a standard username & password popup that is default of the browser they use.

The important setting on the server side is the [access control allow credentials](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Access-Control-Allow-Credentials)

bmatthieu3 commented:
Thanks, it is ready to be merged
